src/Client.cpp

```diff
@@ -1,245 +1,487 @@
+    
 }
 
-bool Client::respond(const msp::Response &response, const bool wait_ack) {
-    return respond_raw(uint8_t(response.id()), response.encode(), wait_ack);
+
+bool Client::sendMessageNoWait(msp::Message& message)
+{
+    if (print_debug_) std::cout << "async sending message - ID " << (uint32_t)message.id() << std::endl;
+    if(!sendData(message.id(), message.encode())) { 
+        std::cout << "async sendData failed" << std::endl;
+        return false; 
+    }
+    return true;
 }
 
-bool Client::respond_raw(const uint8_t id, const ByteVector &data, const bool wait_ack) {
-    // send response
-    if(!sendData(id, data)) { return false; }
 
-    if(!wait_ack)
-        return true;
+uint8_t Client::extractChar() {
+    if(buffer.sgetc()==EOF) {
+        std::cout << "buffer returned EOF; reading char directly from port" << std::endl;
+        asio::read(port, buffer, asio::transfer_exactly(1));
+    }
+    uint8_t rc = uint8_t(buffer.sbumpc());
+    return rc;
+}
 
-    // wait for thread to received message
-    std::unique_lock<std::mutex> lock(mutex_cv_ack);
-    cv_ack.wait(lock, [&]{
-        mutex_request.lock();
-        const bool received = (request_received.id==id);
-        // unlock to wait for next message
-        if(!received) { mutex_request.unlock(); }
-        return received;
-    });
 
-    // check status, expect ACK without payload
-    const bool success = request_received.status==OK;
-    mutex_request.unlock();
-    return success;
+bool Client::sendData(const msp::ID id, const ByteVector &data) 
+{
+    if (print_debug_) std::cout << "sending: " << uint32_t(id) << " | " << data;
+    ByteVector msg;
+    if (msp_ver_ == 2) {
+        msg = packMessageV2(id,data);
+    } else {
+        msg = packMessageV1(id,data);
+    }
+    if (print_debug_) std::cout << "packed: " << msg;
+    asio::error_code ec;
+    std::size_t bytes_written;
+    {
+        std::lock_guard<std::mutex> lock(mutex_send);
+        bytes_written = asio::write(port, asio::buffer(msg.data(), msg.size()), ec);
+    }
+    if (ec == asio::error::operation_aborted) {
+        //operation_aborted error probably means the client is being closed
+        std::cout << "------------------> WRITE FAILED <--------------------" <<std::endl;
+        return false;
+    }
+    if (print_debug_) std::cout << "write complete: " << bytes_written << " vs " << msg.size() << std::endl;
+    return (bytes_written==msg.size());
+    
+}
+
+
+//TODO: if id is > 254, pack using V2 over V1 protocol
+ByteVector Client::packMessageV1(const msp::ID id, const ByteVector &data) 
+{
+    
+    ByteVector msg;
+    msg.push_back('$');                                 // preamble1
+    msg.push_back('M');                                 // preamble2
+    msg.push_back('<');                                 // direction
+    msg.push_back(uint8_t(data.size()));                // data size
+    msg.push_back(uint8_t(id));                                  // message_id
```

Comment:
In the comments here (message_id) and in two lines (crc) there are too many white spaces/tabs. Align the comments with the lines above.